mlir/lib/Transform/AIRMiscPasses.cpp

```diff
@@ -879,6 +881,119 @@ void AIRLowerHerdParallelPass::runOnOperation() {
+
+  LogicalResult matchAndRewrite(Operation *op,
+                                PatternRewriter &rewriter) const override {
+    auto chanOp = dyn_cast<air::ChannelInterface>(op);
+    if (!chanOp)
+      return failure();
+    MemRefType memrefTy = chanOp.getMemref().getType().cast<MemRefType>();
+
+    // channel op must be in an scf.for loop
+    auto forOp = op->getParentOfType<scf::ForOp>();
+    if (!forOp)
+      return failure();
+
+    // channel op must be the only op in the loop
+    if (forOp.getBody()->getOperations().size() != 2)
+      return failure();
+
+    // scf.for must have constant bounds and step
+    auto lb = forOp.getLowerBound().getDefiningOp<arith::ConstantIndexOp>();
+    auto ub = forOp.getUpperBound().getDefiningOp<arith::ConstantIndexOp>();
+    auto step = forOp.getStep().getDefiningOp<arith::ConstantIndexOp>();
+    if (!lb || !ub || !step)
+      return failure();
+
+    auto trip_count = (ub.value() - lb.value()) / step.value();
+
+    rewriter.setInsertionPoint(forOp);
+    auto loc = op->getLoc();
+
+    SmallVector<Value> offsets;
+    Value iv = nullptr;
+    int ivOffsetIdx = 0;
+    for (auto v : chanOp.getOffsets()) {
+      auto c = dyn_cast_if_present<arith::ConstantIndexOp>(v.getDefiningOp());
+      if (!c) {
+        if (v == forOp.getInductionVar()) {
+          if (iv) // there can be only one
+            return failure();
+          iv = v;
+          v = rewriter.create<arith::ConstantIndexOp>(loc, 0);
+        }
+      }
+      offsets.push_back(v);
+      if (!iv)
+        ivOffsetIdx++;
+    }
+
+    SmallVector<Value> sizes;
+    sizes.push_back(rewriter.create<arith::ConstantIndexOp>(loc, trip_count));
+    for (auto v : chanOp.getSizes()) {
+      auto c = dyn_cast<arith::ConstantIndexOp>(v.getDefiningOp());
+      if (!c)
+        return failure();
+      sizes.push_back(c);
+    }
+
+    SmallVector<Value> strides;
+    auto lastStride = dyn_cast<arith::ConstantIndexOp>(
+        chanOp.getStrides()[ivOffsetIdx].getDefiningOp());
+    auto foldedStride = step.value() * lastStride.value();
+
+    strides.push_back(
+        rewriter.create<arith::ConstantIndexOp>(loc, foldedStride));
+    for (auto v : chanOp.getStrides()) {
+      auto c = dyn_cast<arith::ConstantIndexOp>(v.getDefiningOp());
+      if (!c)
+        return failure();
+      strides.push_back(c);
+    }
+
+    SmallVector<Value> deps;
+    SmallVector<Type> tys;
+    auto ctx = op->getContext();
+    auto channel = FlatSymbolRefAttr::get(ctx, chanOp.getChanName());
+    if (isa<air::ChannelPutOp>(op))
+      rewriter.create<air::ChannelPutOp>(
+          loc, tys, deps, channel, chanOp.getIndices(), chanOp.getMemref(),
+          offsets, sizes, strides);
+    else
+      rewriter.create<air::ChannelGetOp>(
+          loc, tys, deps, channel, chanOp.getIndices(), chanOp.getMemref(),
+          offsets, sizes, strides);
+    rewriter.eraseOp(op);
+    rewriter.eraseOp(forOp);
+    return success();
+  }
+};
```

Comment:
Ouch, 85 lines of mostly uncommented code... Feel free to help some eventual readers. :-)